zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/multiinstance/MultiInstanceBatchedSubProcessesTest.java

```diff
@@ -0,0 +1,401 @@
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.zeebe.engine.processing.bpmn.multiinstance;
+
+import io.camunda.zeebe.engine.util.EngineRule;
+import io.camunda.zeebe.model.bpmn.Bpmn;
+import io.camunda.zeebe.model.bpmn.BpmnModelInstance;
+import io.camunda.zeebe.protocol.record.intent.ProcessInstanceIntent;
+import io.camunda.zeebe.test.util.record.RecordingExporter;
+import io.camunda.zeebe.test.util.record.RecordingExporterTestWatcher;
+import org.assertj.core.api.Assertions;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class MultiInstanceBatchedSubProcessesTest {
+
+  @ClassRule
+  public static final EngineRule ENGINE = EngineRule.singlePartition().maxCommandsInBatch(47);
+
+  private static final String PROCESS_ID = "process";
+  private static final String SUBPROCESS_2_START_EVENT = "subprocess2StartEvent";
+  private static final BpmnModelInstance PROCESS =
+      Bpmn.createExecutableProcess(PROCESS_ID)
+          .startEvent()
+          .zeebeOutputExpression(getProcessVariable(), "bill")
+          .subProcess(
+              "subprocess1",
+              s ->
+                  s.multiInstance(
+                      b ->
+                          b.zeebeInputCollectionExpression("bill.lines").zeebeInputElement("line")))
+          .embeddedSubProcess()
+          .startEvent("sub-process-start")
+          .subProcess(
+              "subprocess2",
+              s ->
+                  s.multiInstance(
+                      b ->
+                          b.zeebeInputCollectionExpression("line.orders")
+                              .zeebeInputElement("order")))
+          .embeddedSubProcess()
+          .startEvent(SUBPROCESS_2_START_EVENT)
+          .exclusiveGateway("exclusive-gateway")
+          .defaultFlow()
+          .exclusiveGateway()
+          .endEvent()
+          .subProcessDone()
+          .endEvent()
+          .subProcessDone()
+          .endEvent()
+          .done();
+
+  @Rule
+  public final RecordingExporterTestWatcher recordingExporterTestWatcher =
+      new RecordingExporterTestWatcher();
+
+  @Test
+  public void shouldCompleteAllChildInstancesBeforeCompletingTheMultiInstanceBody() {
+    // given
+    ENGINE.deployment().withXmlResource(PROCESS).deploy();
+    final int orderCount = 50;
+
+    // when
+    ENGINE.processInstance().ofBpmnProcessId(PROCESS_ID).create();
+
+    // then
+    Assertions.assertThat(
+            RecordingExporter.processInstanceRecords(ProcessInstanceIntent.ELEMENT_ACTIVATED)
+                .withElementId(SUBPROCESS_2_START_EVENT)
+                .limit(orderCount)
+                .toList())
+        .hasSize(orderCount);
+  }
+
+  private static String getProcessVariable() {
```

Comment:
❓ Do you think you could simplify this test case now that we understand the real problem? I believe we should able to reproduce with a single (non-nested) multi-instance subprocess with an input collection of only 2 elements 🙂

Reply:
I actually tried simplifying to `1 non-nested MI with 5 items` and `nested multi instance of two lines each with two orders` and adjust the batch size, and in both cases I failed to reproduce.

Which probably says it has something to do with the size of the input collection too 😕 🤔

Reply:
Try this one 😉 

```java
/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Camunda License 1.0. You may not use this file
 * except in compliance with the Camunda License 1.0.
 */
package io.camunda.zeebe.engine.processing.bpmn.multiinstance;

import io.camunda.zeebe.engine.util.EngineRule;
import io.camunda.zeebe.model.bpmn.Bpmn;
import io.camunda.zeebe.model.bpmn.BpmnModelInstance;
import io.camunda.zeebe.protocol.record.intent.ProcessInstanceIntent;
import io.camunda.zeebe.test.util.record.RecordingExporter;
import io.camunda.zeebe.test.util.record.RecordingExporterTestWatcher;
import org.assertj.core.api.Assertions;
import org.junit.ClassRule;
import org.junit.Rule;
import org.junit.Test;

public class MultiInstanceBatchedSubProcessesTest {

  @ClassRule
  public static final EngineRule ENGINE = EngineRule.singlePartition().maxCommandsInBatch(6);

  private static final String PROCESS_ID = "process";
  private static final BpmnModelInstance PROCESS =
      Bpmn.createExecutableProcess(PROCESS_ID)
          .startEvent()
          .zeebeOutputExpression("= [1,2]", "inputCollection")
          .subProcess(
              "subprocess1",
              s ->
                  s.multiInstance(
                      b ->
                          b.zeebeInputCollectionExpression("inputCollection")
                              .zeebeInputElement("input")))
          .embeddedSubProcess()
          .startEvent("sub-process-start")
          .subProcessDone()
          .endEvent()
          .done();

  @Rule
  public final RecordingExporterTestWatcher recordingExporterTestWatcher =
      new RecordingExporterTestWatcher();

  @Test
  public void shouldCompleteAllChildInstancesBeforeCompletingTheMultiInstanceBody() {
    // given
    ENGINE.deployment().withXmlResource(PROCESS).deploy();
    final int count = 2;

    // when
    ENGINE.processInstance().ofBpmnProcessId(PROCESS_ID).create();

    // then
    Assertions.assertThat(
            RecordingExporter.processInstanceRecords(ProcessInstanceIntent.ELEMENT_ACTIVATED)
                .withElementId("sub-process-start")
                .limit(count)
                .toList())
        .hasSize(count);
  }
}

```

The reason you couldn't reproduce it is because the amount of commands processed in the subprocess exceeded the maximum command batch size. Because of that the subprocess didn't complete fully. By reducing the amount of commands written in the subprocess we can reproduce it with a more minimal example 🙂 

I'd also suggest to link the issue in the test case to specify it's a regression test for this bug. That gives some more context to understand it later on.

Reply:
Thank you! 🙇